app/models/qdm/basetypes/data_element.rb

```diff
@@ -89,6 +89,53 @@ def shift_facility_location_dates(facility_location, seconds)
       facility_location['locationPeriod'][:high] = (facility_location['locationPeriod'][:high].to_time + seconds).to_datetime
     end
 
+    # The necessary reason for this function is to avoid a problem when shifting
+    # past a year that is a leap year. February 29th dates are handled by moving
+    # back to the 28th in non leap years
+    def shift_years(year) 
+      # Iterate over fields
+      fields.keys.each do |field|
+        # Check if field is a DateTime
+        if send(field).is_a? DateTime
+          if send(field).month == 2 && send(field).day == 29 && !Date.leap?(year.to_i)
+            send(field + '=', send(field).change(year: year.to_i, day: 28))
```

Comment:
This is setting the year on the data element to the year passed in, which is passed in like this from Bonnie:

```
  def shift_years(measure, year)
    CQM::Patient.by_user_and_hqmf_set_id(current_user, measure.hqmf_set_id).each do |patient|
      patient.qdmPatient.dataElements.each do |data_element|
        data_element.shift_years(year)
      end
      patient.save!
    end
  end
```

Instead, it should shift the years by the difference between the data element's current year and the old measurement period.  We can calculate that in the `measurement_period` function in Bonnie before updating it on the measure, then pass it into this function.

This function currently will only be accurate if the year was the same as the old measurement period.

This function should instead take a parameter that says how many years to shift by (positive or negative), and then adjust the dates by adding or subtracting that value from the data element's current year.

Reply:
Ah yes, that is my bad, I will adjust that now.